cmd/storage-rest-server.go

```diff
@@ -787,6 +794,161 @@ func waitForHTTPResponse(respBody io.Reader) (io.Reader, error) {
 	}
 }
 
+// drainCloser can be used for wrapping an http response.
+// It will drain the body before closing.
+type drainCloser struct {
+	rc io.ReadCloser
+}
+
+// Read forwards the read operation.
+func (f drainCloser) Read(p []byte) (n int, err error) {
+	return f.rc.Read(p)
+}
+
+// Close drains the body and closes the upstream.
+func (f drainCloser) Close() error {
+	xhttp.DrainBody(f.rc)
+	return nil
+}
+
+// httpStreamResponse allows streaming a response, but still send an error.
+type httpStreamResponse struct {
+	done  chan error
+	block chan []byte
+	err   error
+}
+
+// Write part of the the streaming response.
+// Note that upstream errors are currently not forwarded, but may be in the future.
+func (h *httpStreamResponse) Write(b []byte) (int, error) {
+	tmp := make([]byte, len(b))
+	copy(tmp, b)
+	h.block <- tmp
+	return len(b), h.err
+}
+
+// CloseWithError will close the stream and return the specified error.
+// This can be done several times, but only the first error will be sent.
+// After calling this the stream should not be written to.
+func (h *httpStreamResponse) CloseWithError(err error) {
+	if h.done == nil {
+		return
+	}
+	h.done <- err
+	h.err = err
+	// Indicates that the response is done.
+	<-h.done
+	h.done = nil
+}
+
+// streamHTTPResponse can be used to avoid timeouts with long storage
+// operations, such as bitrot verification or data usage crawling.
+// Every 10 seconds a space character is sent.
+// The returned function should always be called to release resources.
+// An optional error can be sent which will be picked as text only error,
+// without its original type by the receiver.
+// waitForHTTPStream should be used to the receiving side.
+func streamHTTPResponse(w http.ResponseWriter) *httpStreamResponse {
```

Comment:
Why does `keepalive` not serve the purpose of keeping the connection active?

Reply:
Great Q. Let's look at it from the callers position.

This is how it is used:
```
	writer := streamHTTPResponse(w)
	writer.CloseWithError(s.storage.WalkDir(r.Context(), WalkDirOptions{...}, writer))
```
Using the `keepHTTPResponseAlive` it would look something like this:
```
	done := keepHTTPResponseAlive(w)
	done(nil)
	err := s.storage.WalkDir(r.Context(), WalkDirOptions{Bucket: volume, BaseDir: dirPath, Recursive: recursive}, w)
````

So, first of all `keepHTTPResponseAlive` is not really doing anything. Since we are streaming to the response writer `w` we need to close it right away so we don't race, making it useless.

We can do something else:
```
	done := keepHTTPResponseAlive(w)
	var tmp bytes.Buffer
	err := s.storage.WalkDir(r.Context(), WalkDirOptions{...}, &tmp)
	if err == io.EOF {
		err = nil
	}
	done(err)
	if err == nil {
		w.Write(tmp.Bytes())
	}
```
First of all this is not streaming anymore since we collect everything in a buffer. If the call (as here) can take several minutes and collect huge result this is bad in both ways.

Another thing is that we can not distinguish between `io.EOF` and nil.

Compare this to the cleanliness of this:

```
	writer := streamHTTPResponse(w)
	writer.CloseWithError(s.storage.WalkDir(r.Context(), WalkDirOptions{...}, writer))
```

Fully streaming, with keepalive padding and allows to return an error at any point in the stream.